pkg/api/validate/admin_test.go

```diff
@@ -43,8 +42,8 @@ func TestAdminAPIValidateUpdate(t *testing.T) {
 			f: func(oc *api.OpenShiftManagedCluster) {
 				oc.Config.PluginVersion = "latest" // the RP does this, but after validation: the user can't do this
 			},
-			expectedErrs: []error{
-				errors.New(`invalid change [Config.PluginVersion: latest != ]`),
+			expectedErrs: []*regexp.Regexp{
+				regexp.MustCompile(`PluginVersion:\s+?"latest"(?s).+?PluginVersion:\s+?""`),
```

Comment:
Output of `google/go-cmp` is unstable and the library [randomly adds non-breaking space and a regular space](https://github.com/google/go-cmp/blob/1b316004397f1f336546ca058ddb5b95c41a8772/cmp/report_text.go#L21-L49) into the output, so we can't just copy-paste the output and use it in tests: we have to use a regexp.

Reply:
these are really hard to read. I propose using a custom diff reporter for our go-cmp wrapper. The following is code which will produce such lists of differences in our tests:

```
admin_test.go:116: received errors:
admin_test.go:118: 	invalid change [{*api.OpenShiftManagedCluster}.Properties.AgentPoolProfiles[2].Count: 4 != 1]
admin_test.go:111: test case "invalid change"
admin_test.go:112: expected errors:
admin_test.go:114: 	invalid change [Name: new != openshift]
admin_test.go:116: received errors:
admin_test.go:118: 	invalid change [{*api.OpenShiftManagedCluster}.Name: new != openshift]
```

The code:

```
package cmp

import (
	"crypto/x509"
	"fmt"
	"strings"

	gocmp "github.com/google/go-cmp/cmp"
)

// simpleDiffReporter is a simple diff reporter that only records differences
// detected during comparison.
type simpleDiffReporter struct {
	path  gocmp.Path
	diffs []string
}

func (r *simpleDiffReporter) PushStep(ps gocmp.PathStep) {
	r.path = append(r.path, ps)
}

func (r *simpleDiffReporter) Report(rs gocmp.Result) {
	if !rs.Equal() {
		vx, vy := r.path.Last().Values()
		r.diffs = append(r.diffs, fmt.Sprintf("[%+v: %+v != %+v]", r.path.GoString(), vx, vy))
	}
}

func (r *simpleDiffReporter) PopStep() {
	r.path = r.path[:len(r.path)-1]
}

func (r *simpleDiffReporter) String() string {
	return strings.Join(r.diffs, "\n")
}

// Diff is a wrapper for github.com/google/go-cmp/cmp.Diff using the custom
// simple diff reporter
func Diff(x, y interface{}) string {
	// TODO(mikalai): Remove x509CertComparer after upgrading to a Go version
	//  that includes https://github.com/golang/go/issues/28743
	var r simpleDiffReporter
	if !gocmp.Equal(x, y, gocmp.Reporter(&r), gocmp.Comparer(x509CertComparer)) {
		return r.String()
	}
	return ""
}

func x509CertComparer(x, y *x509.Certificate) bool {
	if x == nil || y == nil {
		return x == y
	}
	return x.Equal(y)
}
```

You just need to change the diff strings in the tests and no need to use regexes

Reply:
@charlesakalugwu This is very similar (almost exactly the same) to what I had while working on this PR. I decided not to go with a custom reporter because:

* I didn't want to introduce an extra unit in our source code only to improve readability of tests. We will have to write tests and maintain this custom reporter.
* I wanted us to be able to just remove `pkg/util/cmp` after upgrading to a new version of Go and use `google/go-cmp` as is.

I agree that regexp is not very readable, but it's not too bad. Also test is not something one will be reading regularly. Someone will have to read unit tests when changing/breaking something in the code we test. In this case the person will know what they are doing and it will not be a problem to read the regexp.

Do you think improving test readability worth introducing new code?

We can simplify expressions a bit. We can check only for a line with a new value and ignore the old value. For example, we can replace ``regexp.MustCompile(`PluginVersion:\s+?"latest"(?s).+?PluginVersion:\s+?""`)`` with ``regexp.MustCompile(`PluginVersion:\s+?"latest"`)``

What do you think?